samza-api/src/main/java/org/apache/samza/checkpoint/CheckpointListener.java

```diff
@@ -22,11 +22,28 @@
 import org.apache.samza.system.SystemStreamPartition;
 
 /**
- * This interface should be implemented by SystemConsumers which want to be notified about checkpoint values
- * of the most recent checkpoint.
- * Note: For broadcast streams, different tasks may register the same ssp with different values. It
- * is the callee's responsibility to select the lowest one.
+ * SystemConsumers that want to be notified about checkpoints for their SSPs and potentially modify them may
+ * implement this interface.
+ *
+ * Note: For broadcast streams, different Tasks may checkpoint the same ssp with different offsets. It is the
+ * SystemConsumer's responsibility to select the appropriate (e.g., lowest) offset for its own use.
  */
 public interface CheckpointListener {
-  void onCheckpoint(Map<SystemStreamPartition, String> offsets);
+  /**
+   * Called before writing the checkpoint for each Task in the Container processing SSPs on this system.
+   * Implementations may modify the Task offsets to be checkpointed, or return the provided offsets as-is.
+   *
+   * @param offsets offsets about to be checkpointed.
+   * @return potentially modified offsets to be checkpointed.
+   */
+  default Map<SystemStreamPartition, String> beforeCheckpoint(Map<SystemStreamPartition, String> offsets) {
```

Comment:
Please correct me if i'm wrong here.

1. Why do we need this API? When does the offsets tracked by the TaskInstance/RunLoop  for a stream has to be invalidated/altered by an external listener API?  

2. Would it make sense to add an API in systemAdmin like getCommitableOffsets/getValidOffsets which would return the valid offsets based upon the provided input offsets that we can then commit. 

In samza, checkpoint listeners are the system consumers which implement the CheckpointListener interface in addition to SystemConsumer interface. The problem with adding this API in checkpointListener is that every systemconsumer now has to be aware that it had to implement CheckpointListener to have an option to manipulate checkpoints before commit.

Reply:
> Why do we need this API?
See earlier discussion. This is a port for the LI internal logic for getting safe offset from KafkaConsumer for large message support.

Checkpoint listener interface already exists, which gives consumers a callback after the checkpoint has been written. We need a new callback before the checkpoint is written with the ability to modify it. It makes sense to me to put this callback int the CheckpointListener as well. WDYT?

SystemConsumer is typically the one that returns/mints the offsets for incoming messages, so it's seems reasonable to give the checkpoint callback to the SystemConsumer, instead of proxying to the same SystemConsumer through the SystemAdmin. WDYT?

Reply:
Makes sense. Thanks for the explanation.